proposals/SelectorResourceBudgets.md

```diff
@@ -0,0 +1,207 @@
+	- This is a sanity-checking process for the either design phase.
+4. Implement: in the [go-ipld-prime/traversal/selectors](https://github.com/ipld/go-ipld-prime/tree/master/traversal/selector) package.
+5. Test: make sure we have examples of datasets and selectors to run on them which we expect to be halted by budget limits.
+	- Ideally this should be in language-agnostic test fixture files, so we can reuse them in other selectors implementations.
+6. Documentation: update it.
+7. Synchronize: other implementations!
+	- The [ChainSafe forest](https://github.com/ChainSafe/forest/) project contains a Selectors implementation -- communicate with them about these changes!
+8. Propagation to downstream, possible small migrations?
+	- If we make the budget system non-optional, then existing Selector documents may not work.
+	- Or, there might be no special work needed here, if the budget system is entirely optional.
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+- Selectors in the go-ipld-prime implementation should have resource budgeting.
+- Test fixtures should demonstrate what a selection which halts due to a resource budget exhaustion behaves like.
+- The resource budget specification declaration system should be reasonably comprehensible and look like something we can tell administrators of hypothetical services using this system how to configure.
+	- Probably: it should be as simple as _one number_.
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+When developers such as the Filecoin team feel comfortable exposing features using Selectors to users, then this project is a success.
+
+#### Counterpoints & pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+- Overcomplicating the budget system could result in usability failure.
+  (Arguably, the current limit systems are this, because they're too granular, which is no substitute for a holistic system.)
+
+- Technical consideration: Beware the "[Billion Laughs](https://en.wikipedia.org/wiki/Billion_laughs)" problem.
+  (This is why this document keeps emphasizing a budget that is holistic and monotonically decreasing.)
+
+- A system that halts but returns insufficient information about why could be frustrating to users,
+  even if it successfully addresses the DoS problem.
+
+- Keep in mind: this proposal only describes implementing this in golang.
+  We do not currently have a javascript Selectors implementation (and creating one is a larger task).
+  This is not a problem per se; it's just something to remember when considering what can be immediately built upon this work.
+
+#### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
```

Comment:
I assume, but is there any concept of pagination/partial results that can be applied for selectors?

Reply:
It has been discussed, but no such thing is implemented nor shipped at present.

Resuming that discussion would probably be a part of the work that would go on while engaging on this project.

Reply:
In practice, it's my understanding that at present, systems using Selectors get in the habit of launching small queries (depth limited, or, constructed to favor left-leaning trees for example) to get started with exploring the data, and use more queries subsequently.

This "works" but obviously leaves some load to the brain of the human crafting the Selector, which isn't really the most desired outcome.

Reply:
Some prior discussion involved ideas like "what if we could ask the selector to return every (N % 3 == 1) blocks?" and similar ideas.  The aim there was to end up with something that you could imagine a system generating automatically in order to fan out queries for data to multiple peers and start getting different fractions of the data back from them in parallel.

This only got to the discussion phase.  There may be neat ideas here, but they trend towards getting complicated, so we pushed them out of the first round of Selector work.